cifs/parse_cifs.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cifs
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"regexp"
+	"strconv"
+	"strings"
+)
+
+// ParseClientStats returns stats read from /proc/fs/cifs/Stats
+func ParseClientStats(r io.Reader) (*ClientStats, error) {
+	stats := &ClientStats{}
+	stats.Header = make(map[string]uint64)
+	scanner := bufio.NewScanner(r)
+	// Parse header
+	for scanner.Scan() {
+		line := scanner.Text()
+		for _, regexpHeader := range regexpHeaders {
+			match := regexpHeader.FindStringSubmatch(line)
+			if 0 == len(match) {
+				continue
+			}
+			for index, name := range regexpHeader.SubexpNames() {
+				if 0 == index || "" == name {
+					continue
+				}
+				value, err := strconv.ParseUint(match[index], 10, 64)
+				if nil != err {
+					continue
+				}
+				stats.Header[name] = value
+			}
+			break
+		}
+		if strings.HasPrefix(line, "Total vfs") {
+			break
+		}
+	}
+	// Parse Shares
+	var tmpSMB1Stats *SMB1Stats
+	var tmpSMB2Stats *SMB2Stats
+	var tmpSessionIDs *SessionIDs
+	// The legacy variable sets the current context. True for SMB1, False for SMB2
+	legacy := true
+	for scanner.Scan() {
+		line := scanner.Text()
+		if legacy {
+			// This part manages the parsing of SMB1
+			for _, regexpSMB1 := range regexpSMB1s {
+				match := regexpSMB1.FindStringSubmatch(line)
+				if 0 == len(match) {
```

Comment:
I understand this is subjective, but I really am not a fan of comparisons with a constant on the left side as opposed to the right.  It is quite uncommon in Go.